core/camel-support/src/main/java/org/apache/camel/support/MessageSupport.java

```diff
@@ -306,4 +310,20 @@ protected String createMessageId() {
         }
     }
 
+    @Override
+    public boolean hasTrait(MessageTrait trait) {
+        Object payload = traits[trait.ordinal()];
+
+        return payload != null;
+    }
+
+    @Override
+    public Object getPayloadForTrait(MessageTrait trait) {
+        return traits[trait.ordinal()];
+    }
+
+    @Override
+    public void setPayloadForTrait(MessageTrait trait, Object object) {
+        traits[trait.ordinal()] = object;
+    }
```

Comment:
This part here: if accepted, then I'll do create a separate class to abstract this. We already do a similar thing when handling the `internalProperties` on the [AbstractExchange](https://github.com/apache/camel/blob/main/core/camel-support/src/main/java/org/apache/camel/support/AbstractExchange.java#L252-L276).